### ISSUE TITLE
Trouble replicating results from Optuna trial

### ISSUE THREAD

**Issue by maxmatical** [optuna/optuna#975]
I've been trying to use Optuna with fastai. When I run the study, I get the following values

```
[I 2020-03-03 19:37:32,469] Finished trial#44 resulted in value: 0.5143198370933533. Current best value is 0.8752983212471008 with parameters: {'lr': 6.788371449432635e-05, 'pct_start': 0.16117056838283858, 'b1': 0.8702881830011164, 'b2': 0.8851403535800838, 'eps': 0.001, 'wd': 0.0007973653050235285}.
```

However, if I create the learner with the same hyperparameters, I cannot reproduce 0.875, and the accuracy is stuck at 0.5143198370933533.

<img width="413" alt="image" src="https://user-images.githubusercontent.com/8890262/75813543-26f32780-5d5e-11ea-9f9b-34e349b67c47.png">

**Comment by maxmatical** [comment]
I am using a custom model (from huggingface transformers) with fastai. It it possible that the model weights aren't being reset across different trials? Because it seems that the trial result values always seem to converge to a specific value (0.5143198370933533)

**Comment by toshihikoyanase** [comment]
Hi @maxmatical,

>  the model weights aren't being reset across different trials?

Hmm. I'm not familiar with fastai, but I couldn't find any code to reset the weights/states shared with other trials in [`examples/fastai_simple.py`](https://github.com/optuna/optuna/blob/master/examples/fastai_simple.py). I think more information about your code is required for further troubleshooting. So, it will be a great help if you share your code (or examples with the characteristics of your code).

BTW, the reproducibility of training is a common issue of the hyperparameter optimization of the neural networks. One possible approach is to save the models trained in objective functions instead of recreating the model after the optimization. You can find further information including example code in the following document:
https://optuna.readthedocs.io/en/stable/faq.html#how-to-save-machine-learning-models-trained-in-objective-functions.

**Comment by maxmatical** [comment]
The code looks like this

```
def objective(trial):
    lr = trial.suggest_loguniform('lr', 1e-6, 1e-3)
    pct_start = trial.suggest_uniform('pct_start', 0.05, 0.5)
    b1 = trial.suggest_uniform('b1', 0.7, 0.9)
    b2 = trial.suggest_uniform('b2', 0.6, 0.98)
    eps = trial.suggest_categorical('eps', [1e-1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6])
    wd = trial.suggest_loguniform('wd', 1e-8, 1e-3)

    # destroy old learner
    try: 
        learn.destroy()
    except:
        print('no learner created')

    learn = Learner(data_clas,
        custom_transformer_model,
        opt_func = lambda input: AdamW(input,correct_bias=False, eps=eps),
        loss_func = FlattenedLoss(LabelSmoothingCrossEntropy, axis=-1),
        metrics = [accuracy],
        wd = wd,
        callback_fns=[partial(FastAIPruningCallback, trial=trial, monitor='accuracy')])
    # For roberta-base
    list_layers = [learn.model.transformer.roberta.embeddings,
                learn.model.transformer.roberta.encoder.layer[0],
                learn.model.transformer.roberta.encoder.layer[1],
                learn.model.transformer.roberta.encoder.layer[2],
                learn.model.transformer.roberta.encoder.layer[3],
                learn.model.transformer.roberta.encoder.layer[4],
                learn.model.transformer.roberta.encoder.layer[5],
                learn.model.transformer.roberta.encoder.layer[6],
                learn.model.transformer.roberta.encoder.layer[7],
                learn.model.transformer.roberta.encoder.layer[8],
                learn.model.transformer.roberta.encoder.layer[9],
                learn.model.transformer.roberta.encoder.layer[10],
                learn.model.transformer.roberta.encoder.layer[11],
                learn.model.transformer.roberta.pooler]

    learn.split(list_layers)
    learn.load('initial')
    learn.unfreeze() 
    learn.to_fp16()
    learn.fit_one_cycle(1,
                lr,
                pct_start = pct_start,
                moms = (b1, b2))

    return learn.validate()[-1].item() # returns accuracy

```
Where `custom_transformer_model` is loading a pre-trained ROBERTA model. I even made sure to check that the fastai learner was being destroyed at the end of each trial. When I print out the model weights at the before training in each trial, I get something like
<img width="1087" alt="image" src="https://user-images.githubusercontent.com/8890262/75877511-e9d27800-5de5-11ea-84da-8848f606cb47.png">

However, if I save the initial model weights beforehand and load them at the beginning of each trial with `learn.load('initial')`, the weights remain consistent at each trial

<img width="901" alt="image" src="https://user-images.githubusercontent.com/8890262/75877775-6a917400-5de6-11ea-9575-78685a589054.png">

**Comment by ofersabo** [comment]
I believe I encountered the same issue. 

I'm running distributed experiments across servers. When I try to reproduce the best experiment, I copy all the parameters from the config file (I'm using the optuna with allennlp) including seed, and for some reason, I can't reproduce this exact same results ( I also run the same experiment on the server).
reproduction is fine when I don't use the optuna distributed trial, Also it doesn't happen when I use my own model, i.e creating my own pytorch model and experiment it with optuna distributed trial. 
It happens when I load a pre-trained model from huggingface.  
Any ideas why this may happen? 

I'll try to see if the model parameters are different at the beginning of each trial.
